Consolidate time series downsampling docs (#2274)
elastic/docs-content: manage-data/data-store/data-streams/run-downsampling.md

---
applies_to:
  stack: ga
  serverless: ga
navigation_title: "Downsample data"
mapped_pages:
  - https://www.elastic.co/guide/en/elasticsearch/reference/current/downsampling-manual.html
  - https://www.elastic.co/guide/en/elasticsearch/reference/current/downsampling-ilm.html
products:
  - id: elasticsearch
---

# Downsample time series data [running-downsampling]

To downsample a time series data stream (TSDS), you can use index lifecycle management (ILM) or a data stream lifecycle. (You can also use the [downsample API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-indices-downsample) with an individual time series index, but most users don't need to use the API.)

Before you begin, review the [](downsampling-concepts.md).

:::{important}
Downsampling requires **read-only** data.
:::

In most cases, you can choose the data stream lifecycle option. If you're using [data tiers](/manage-data/lifecycle/data-tiers.md) in {{stack}}, choose the index lifecycle option.

::::{tab-set}

:::{tab-item} Data stream lifecycle

## Downsample with a data stream lifecycle
```{applies_to}
stack: ga
serverless: ga
```

> To downsample a time series via a data stream lifecycle, add a [downsampling](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-indices-put-data-lifecycle) section to the lifecycle. 

Author: Fixed in d5af0d309ff414c26eca7260bb52f5cd04aa1211

---
applies_to:
  stack: ga
  serverless: ga
navigation_title: "Run downsampling"
mapped_pages:
  - https://www.elastic.co/guide/en/elasticsearch/reference/current/downsampling-manual.html
  - https://www.elastic.co/guide/en/elasticsearch/reference/current/downsampling-ilm.html
products:
  - id: elasticsearch
---

# Run downsampling on time series data [running-downsampling]

:::{admonition} Page status
🟢 Ready for review
:::

% TODO consider retitling (cf. overview)

To downsample a time series data stream backing index, you can use the `downsample API`, index lifecycle management (ILM), or a data stream lifecycle.

:::{note}
Downsampling runs on the data stream backing index, not the data stream itself.
:::

## Prerequisites

Before you start, make sure your index is a candidate for downsampling:

> * The index must be **read-only**. You can roll over a write index and make it read-only.

Author: I tried to address some of this, but please take another look -- not sure I'm following your comments.

To downsample time series data as part of index lifecycle management (ILM), include a [downsample action](elasticsearch://reference/elasticsearch/index-lifecycle-actions/ilm-downsample.md) in your ILM policy:

```console
PUT _ilm/policy/my_policy
{
"policy": {
    "phases": {
    "warm": {
        "actions": {
        "downsample" : {
            "fixed_interval": "1h"
        }
        }
    }
    }
}
}
```
Set `fixed_interval` to your preferred level of granularity. The original time series data will be aggregated at this interval.

% TODO consider restoring removed tutorial-esque content

In this example, an ILM policy is configured for the `hot` phase. The downsample action runs after the index is rolled over and the [index time series end time](elasticsearch://reference/elasticsearch/index-settings/time-series.md#index-time-series-end-time) has passed. 

```console
PUT _ilm/policy/datastream_policy
{
  "policy": {
    "phases": {
      "hot": {
        "actions": {
          "rollover" : {
            "max_age": "5m"
          },
          "downsample": {
  	        "fixed_interval": "1h"
  	      }
        }
      }
    }
  }
}
```

:::

:::{tab-item} Data stream lifecycle

> ## Downsampling with data stream lifecycle management

Reviewer: Unless it changed recently the feature is called `data stream lifecycle` not `data stream lifecycle management` or `data lifecycle management`. Leaving this as a note to ensure the docs mentions this properly.

> To downsample time series data as part of data lifecycle management, create an index template that includes a `lifecycle` section with a [downsampling](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-indices-put-data-lifecycle) object. 

Reviewer: This sentence is confusing different concepts:

-  data lifecycle management refers to both ILM & data stream lifecycle
- It suggests to add the lifecycle section in an index template but the docs it refers to is of the put lifecycle API which has nothing to do with the index template. 

In ILM we suggest the user to add downsampling to their current policy, should we suggest the same here? Add downsampling to the lifecycle, and assume the user knows what we are talking about?

```suggestion
To downsample a TSDS using data stream lifecycle, you can add a downsampling section in the lifecycle. 
```

* Set `fixed_interval` to your preferred level of granularity. The original time series data will be aggregated at this interval.

> * Set `after` to the minimum time to wait after an index rollover, before running downsampling.

Reviewer: ```suggestion
* Set `after` to the minimum time to wait after a backing index rolled over, before running downsampling.
```

Author: Didn't make this change

```console

> PUT _index_template/datastream_template
> {
>   "index_patterns": [
>     "datastream*"
>   ],
>   "data_stream": {},
>   "template": {
>     "lifecycle": {
>       "downsampling": [
>         {
>           "after": "1m",
>           "fixed_interval": "1h"
>         }
>       ]
>     },
>     "settings": {
>       "index": {
>         "mode": "time_series"
>       }
>     },
>     "mappings": {
>       "properties": {
>         "@timestamp": {
>           "type": "date"
>         },
>         [...]
>       }
>     }
>   }
> }

Reviewer: If we are talking about existing TSDS the following could be enough assuming the data stream is already managed by data stream lifecycle 
```suggestion
PUT _data_stream/my-data-stream/_lifecycle
{
  "data_retention": "7d",
  "downsampling": [
     {
       "after": "1m",
       "fixed_interval": "10m"
      },
      {
        "after": "1d",
        "fixed_interval": "1h"
      }
   ]
}
```
What do you think?

Author: Made this change, thanks!

```

> For more details about index templates for time series data streams, refer to [](set-up-tsds.md).

Reviewer: This should be applicable for both ILM & data stream lifecycle, right?

Author: Moved this down, thanks
